packages/core/src/utils/make-namespaced.ts

```diff
@@ -41,55 +40,112 @@ export function makeNamespacedInput(input: SolcInput, output: SolcOutput): SolcI
+                const storageLocation = getStorageLocationAnnotation(contractNode);
+                if (storageLocation !== undefined) {
+                  const structName = contractNode.name;
+                  const variableName = `$${structName}_${(Math.random() * 1e6).toFixed(0)}`;
+                  const insertText = ` ${structName} ${variableName};`;
+
+                  modifications.push(makeInsertAfter(contractNode, insertText));
+                }
+                break;
+              }
+              case 'EnumDefinition':
+              case 'UserDefinedValueTypeDefinition':
+              default: {
+                // - EnumDefinition may be used in structures with storage locations
+                // - UserDefinedValueTypeDefinition may be used in structures with storage locations
+                // - default: in case unexpected ast nodes show up
+                break;
+              }
             }
           }
+          break;
         }
-      } else if (isNodeType('FunctionDefinition', node) || isNodeType('VariableDeclaration', node)) {
-        if (node.documentation) {
-          modifications.push(makeDelete(node.documentation, orig));
+        // - UsingForDirective isn't needed, but it might have NatSpec documentation which is not included in the AST.
+        //   We convert it to a dummy enum to avoid orphaning any possible documentation.
+        case 'UsingForDirective': {
+          const insertText = getUsingForReplacement(node.id);
+          modifications.push(makeReplace(node, orig, insertText));
+          break;
+        }
+        // - ErrorDefinition, FunctionDefinition, and VariableDeclaration might be imported by other files, so they cannot be deleted.
+        //   However, we need to remove their values to avoid referencing other deleted nodes.
+        //   We do this by converting them to dummy variables, but avoiding duplicate names.
+        case 'ErrorDefinition':
+        case 'FunctionDefinition':
+        case 'VariableDeclaration': {
+          // First delete documentation, otherwise it may reference deleted parameters or be orphaned
+          if (node.documentation) {
+            modifications.push(makeDelete(node.documentation, orig));
+          }
+          // If an identifier with the same name was not previously written, replace with a dummy variable.
+          // Otherwise delete to avoid duplicate names, which can happen if there was overloading.
+          // This does not need to check all identifiers from the original contract, since the original compilation
+          // should have failed if there were conflicts in the first place.
+          const name = node.name;
+          if (!replacedIdentifiers.has(name)) {
+            const insertText = `uint256 constant ${name} = 0;`;
+            modifications.push(makeReplace(node, orig, insertText));
+            replacedIdentifiers.add(name);
+          } else {
+            modifications.push(makeDelete(node, orig));
```

Comment:
Why are we deleting and not replacing it with another identifier instead? I saw we're replacing `UsingForDirective` with a dummy enum to avoid orphaning documentation.

Reply:
Is it that variables don't have Natspec?

Reply:
Good point.  We are deleting the associated Natspec here, but there could still be standalone Natspec in the original source code which was not associated with anything.  Added some failing testcases in commit https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/918/commits/a5d1d8815a68296ae860a1aa32d3ef9e45c29da6 to demonstrate, which is the cause of the problem described in https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/917#issuecomment-1828565283.

I think we need to change the overall strategy to avoid deleting anything, but just replace them with dummy enums.  Invalid NatSpec tags seems to have no effect on enums.

Reply:
> but there could still be standalone Natspec in the original source code which was not associated with anything.

This is a great point, thanks for noticing.

> I think we need to change the overall strategy to avoid deleting anything, but just replace them with dummy enums. Invalid NatSpec tags seem to have no effect on enums.

Yeah, I don't think this is trivial, but let's keep it in mind in case NatSpec becomes a problem, which I doubt.

Reply:
Changed it to not delete nodes but replace them with enums in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/918/commits/8badccedb36f428a298397bb1fb699a64071b0b5